contrib: add script for comparing verifier instructions from logs

smagnani96 opened the pull request:
This patch introduces the `verifier_diff.py` script, a utility I find useful when contributing to the datapath. Given two log files containing the output of our BPF verifier tests, this script computes in a CSV and plots the difference between the number of instruction verified within the two runs. It generates plots per-config (bpf_host/1, bpf_host/2, etc.), and in each it reports the insns difference for each verified BPF function (cil_to_netdev, cil_from_netdev, etc.).

The verifier logs can be:

1. taken from specific `ci-verifier` runs: show raw output, copy and save it to a local file, then use the script.
2. generated locally with the same command: `CGO_ENABLED=0 go test -exec "sudo -E" -v -parallel=1 -timeout=20m ./test/verifier -cilium-base-path . -ci-kernel-version 61`

Given two logs `main.log` and `patch.log`, the script can be used as follows:

```bash
$ python3 ./contrib/scripts/verifier_diff.py --file-before main.log --file-after patch.log
2025-06-05 22:46:04,777 [INFO] Output directory insn-diff-patch_log-wrt-main_log-20250605_224604 created.
2025-06-05 22:46:04,777 [INFO] Log files successfully backed up.
2025-06-05 22:46:04,779 [WARNING] Entry ('bpf_sock/1', 'cil_sock_release') missing in patch.log; skipping.
2025-06-05 22:46:04,779 [WARNING] Entry ('bpf_sock/2', 'cil_sock_release') missing in patch.log; skipping.
2025-06-05 22:46:04,783 [INFO] Verifier instructions difference saved as CSV.
2025-06-05 22:46:05,347 [INFO] Saved chart: bpf_host_1.png.
2025-06-05 22:46:05,682 [INFO] Saved chart: bpf_host_2.png.
2025-06-05 22:46:05,928 [INFO] Saved chart: bpf_host_3.png.
2025-06-05 22:46:06,211 [INFO] Saved chart: bpf_host_4.png.
2025-06-05 22:46:06,486 [INFO] Saved chart: bpf_host_5.png.
2025-06-05 22:46:06,755 [INFO] Saved chart: bpf_host_6.png.
2025-06-05 22:46:06,952 [INFO] Saved chart: bpf_overlay_1.png.
2025-06-05 22:46:07,175 [INFO] Saved chart: bpf_overlay_2.png.
2025-06-05 22:46:07,410 [INFO] Saved chart: bpf_xdp_6.png.
2025-06-05 22:46:07,657 [INFO] Saved chart: bpf_xdp_7.png.
2025-06-05 22:46:07,658 [INFO] All charts successfully saved.
```

Please refer to the docstring attached in the script for further details.

I was not sure of PR labels, any adjustments are welcome.

Example of an output plot:

![bpf_xdp_7](https://github.com/user-attachments/assets/03525a08-38cf-4aa5-abd8-d57150cc44ec)

joestringer commented on contrib/scripts/verifier_diff.py:
```diff
@@ -0,0 +1,196 @@
+#!/usr/bin/env python3
```
```suggestion
#!/usr/bin/env python3
# Copyright Authors of Cilium
# SPDX-License-Identifier: Apache-2.0
```

joestringer commented on contrib/scripts/verifier_diff.py:
```diff
@@ -0,0 +1,196 @@
+        description="Compare eBPF number of verified instructions from verifier logs."
+    )
+    parser.add_argument("--file-before", required=True, help="Path to the log before patch")
+    parser.add_argument("--file-after", required=True, help="Path to the log after patch")
```
nit: As a matter of personal preference, I would make one super minor change to just accept the before / after log files as positional parameters to avoid having to type out `--file-before` and `--file-after`. But it's also fine as is.

joestringer commented on contrib/scripts/verifier_diff.py:
I ran `pep8` against this just to see if there was any notable linter complaints, but it was mostly just complaining about line length. I don't think it's meaningful to address those. If we regularly worked in Python in this repository I might add an automated linter, but it seems a bit overkill.

smagnani96 replied:
Yeah, had a different config in my editor 😅 
Should've addressed this in the last force-push, thanks again for it.

joestringer reviewed (approved):
Nice one, thanks for sharing! Just a few super minor nits below.

smagnani96 committed:
contrib: add script for comparing verifier instructions from logs

This patch introduces the `verifier_diff.py` script, a utility I find
useful when contributing to the datapath. Given two log files containing
the output of our BPF verifier tests, this script computes in a CSV and
plots the difference between the number of instruction verified within
the two runs. It generates plots per-config (bpf_host/1, bpf_host/2, etc.),
and in each it reports the insns difference for each verified BPF function
(cil_to_netdev, cil_from_netdev, etc.).

The verifier logs can be:

1. taken from specific `ci-verifier` runs: show raw output, copy and save
   it to a local file, then use the script.
2. generated locally with the same command:
   `CGO_ENABLED=0 go test -exec "sudo -E" -v -parallel=1 -timeout=20m ./test/verifier -cilium-base-path . -ci-kernel-version 61`

Given two logs `main.log` and `patch.log`, the script can be used as follows:

```bash
$ python3 ./contrib/scripts/verifier_diff.py main.log patch.log
2025-06-05 23:57:15,152 [INFO] Output directory insn-diff-patch_log-wrt-main_log-20250605_235715 created.
2025-06-05 23:57:15,152 [INFO] Log files successfully backed up.
2025-06-05 23:57:15,153 [WARNING] Entry ('bpf_sock/1', 'cil_sock_release') missing in patch.log; skipping.
2025-06-05 23:57:15,154 [WARNING] Entry ('bpf_sock/2', 'cil_sock_release') missing in patch.log; skipping.
2025-06-05 23:57:15,155 [INFO] Verifier instructions difference saved as CSV.
2025-06-05 23:57:15,427 [INFO] Saved chart: bpf_host_1.png.
2025-06-05 23:57:15,585 [INFO] Saved chart: bpf_host_2.png.
2025-06-05 23:57:15,707 [INFO] Saved chart: bpf_host_3.png.
2025-06-05 23:57:15,838 [INFO] Saved chart: bpf_host_4.png.
2025-06-05 23:57:15,968 [INFO] Saved chart: bpf_host_5.png.
2025-06-05 23:57:16,095 [INFO] Saved chart: bpf_host_6.png.
2025-06-05 23:57:16,196 [INFO] Saved chart: bpf_overlay_1.png.
2025-06-05 23:57:16,301 [INFO] Saved chart: bpf_overlay_2.png.
2025-06-05 23:57:16,418 [INFO] Saved chart: bpf_xdp_6.png.
2025-06-05 23:57:16,534 [INFO] Saved chart: bpf_xdp_7.png.
2025-06-05 23:57:16,534 [INFO] All charts successfully saved.
```

Please refer to the docstring attached in the script for further details.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>
